plugins/modules/ec2_vpc_vgw.py

```diff
@@ -325,95 +344,92 @@ def find_vgw(client, module, vpn_gateway_id=None):
 
-    return sorted(response["VpnGateways"], key=lambda k: k["VpnGatewayId"])
+    return sorted(response, key=lambda k: k["VpnGatewayId"])
 
 
-def ensure_vgw_present(client, module):
-    # If an existing vgw name and type matches our args, then a match is considered to have been
-    # found and we will not create another vgw.
+def ensure_vgw_present(client, module: AnsibleAWSModule) -> Tuple[bool, Dict[str, Any]]:
+    changed = False
+    params = {
+        "Name": module.params.get("name"),
+        "VpcId": module.params.get("vpc_id"),
+        "VpnGatewayIds": module.params.get("vpn_gateway_id"),
+        "Tags": module.params.get("tags") or {},
+        "PurgeTags": module.params.get("purge_tags", False),
+    }
+
+    existing_vgw = (
+        find_vgw(client, module, [params["VpnGatewayIds"]]) if params["VpnGatewayIds"] else find_vgw(client, module)
+    )
+
+    if existing_vgw:
+        changed |= handle_existing_vgw(client, module, existing_vgw[0], params)
+    else:
+        changed |= create_new_vgw(client, module, params)
+
+    vgw = find_vgw(client, module, [existing_vgw[0]["VpnGatewayId"]]) if existing_vgw else find_vgw(client, module)
+    return changed, get_vgw_info(vgw)
 
+
+def handle_existing_vgw(client, module: AnsibleAWSModule, existing_vgw: dict, params: dict) -> bool:
     changed = False
-    params = dict()
-    result = dict()
-    params["Name"] = module.params.get("name")
-    params["VpcId"] = module.params.get("vpc_id")
-    params["Type"] = module.params.get("type")
-    params["Tags"] = module.params.get("tags")
-    params["VpnGatewayIds"] = module.params.get("vpn_gateway_id")
+    vpn_gateway_id = existing_vgw["VpnGatewayId"]
+
+    # Update tags
+    desired_tags = module.params.get("tags")
+    purge_tags = module.params.get("purge_tags")
+    if desired_tags is None:
+        desired_tags = dict()
+        purge_tags = False
+    tags = dict(Name=module.params.get("name"))
+    tags.update(desired_tags)
+    changed |= ensure_ec2_tags(
+        client, module, vpn_gateway_id, resource_type="vpn-gateway", tags=tags, purge_tags=purge_tags
+    )
 
-    # check that the vpc_id exists. If not, an exception is thrown
+    # Manage VPC attachments
+    current_vpc_attachments = existing_vgw["VpcAttachments"]
     if params["VpcId"]:
-        vpc = find_vpc(client, module)
+        if current_vpc_attachments and current_vpc_attachments[0]["State"] == "attached":
+            current_vpc_id = current_vpc_attachments[0]["VpcId"]
+            if current_vpc_id != params["VpcId"]:
+                if module.check_mode:
+                    return True
+                detach_vgw(client, module, vpn_gateway_id, current_vpc_id)
+                get_waiter(client, "vpn_gateway_detached").wait(VpnGatewayIds=[vpn_gateway_id])
+                attach_vgw(client, module, vpn_gateway_id)
+                changed = True
+        else:
+            attach_vgw(client, module, vpn_gateway_id)
+            changed = True
```

Comment:
```suggestion
        # check first if the VPC is already attached to the VPN
        if any(params["VpcId"] == attachment["VpcId"] for attachment in current_vpc_attachments):
            return changed
        # Now attach the VPC to the Gateway
        if module.check_mode:
            return changed
        attach_vgw(client, module, vpn_gateway_id)
        changed = True
```
The Gateway can be attached to multiple VPC, we should not always detach one attached before attaching another.

Reply:
as far as I understand that was the existing functionality, so did not wanted to change anything that would introduce breaking changes by accident. For now I think it would be okay to move the module to a.aws with same functionality and this issue could be tackled later
@alinabuzachis @abikouo